common/src/main/java/com/tc/net/protocol/transport/ClientMessageTransport.java

```diff
@@ -305,9 +305,7 @@ private void handleSynAck(WireProtocolMessage message) {
           handleHandshakeError(new TransportHandshakeErrorContext(synAck.getErrorContext() + " " + message,
                                                                   synAck.getErrorType()));
         }
-      }
-
-      if (!getConnectionID().isNewConnection() && getConnectionID().isValid()) {
+      } else if (!getConnectionID().isNewConnection() && getConnectionID().isValid()) {
```

Comment:
This worries me a little.  Even if there is a handshake error on reconnect, shouldn't the server not be treating this as a new connection?

Reply:
or is the connection id not properly populated on error

Reply:
Yes probably. My experimentation say this is a pre-existing problem. It's masked in the existing BasicConnection implementation because the thrown AssertionError is in a non TCThreadGroup thread, so it doesn't kill the client. I think the BasicConnection just fires another attempt at the readExecutor and continues on.